### ISSUE TITLE
New transaction + API structure in v4.8.0

### ISSUE THREAD

**Issue by JC5** [firefly-iii/firefly-iii#2194]
@sqrtroot @tonicospinelli @flyingchewbacca @elamperti @Findus23 @nekromoff @evinjon27 @pkoziol @codethief @gavu @4oo4 @8bitgentleman @matiasdelellis @tomazzinko @tornoz @philthynz @techytuppers @emansih @mconway @Ditti4 @filipe-silva @mrahmadt @hamuz @miyoyo @itpony @GeoffreyFrogeye @ogkevin @may-cat @lverissimo @may-cat @mrahmadt @bartonp

Hi everybody,

As you know, Firefly III has supported "split" transactions for quite some time now. The idea is simple: one transaction split in multiple parts. In Firefly III, the database structure surrounding this idea is pretty messy.

This is what a transaction looks like in the database:

![basic-transaction](https://user-images.githubusercontent.com/5889984/54881612-f3233880-4e51-11e9-9a07-e468b8b372ba.png)

The grey thing is a **TransactionJournal** and it holds the red and the green thing (two **Transaction**s) together.

If you create a split transaction, this happens:

![basic-transaction-split](https://user-images.githubusercontent.com/5889984/54881664-6167fb00-4e52-11e9-8729-378edbe19d52.png)

This was a nice solution when I built it but it's hard to maintain: 

1. To match a **Transaction** to it's counterpart I need to compare it to 3, 5, 7 or even more **Transaction**s. I introduced a hack to make this easier but I hate the hack.
2. If the amounts in multiple **Transaction**s are equal, it's very difficult to find the counterpart **Transaction**. Even with the hack.
3. Some views can't handle split transactions, and some can.
4. The API returns inconsistent ID's.
5. I have no idea (beforehand or otherwise) if a transaction is split.
6. Budgets and categories were linked to both the **Transaction** _and_ the **TransactionJournal** because I thought it was a good idea at the time. It wasn't.

So in **Firefly III v4.8.0** I will introduce some backwards incompatible changes that should remedy this. Unfortunately transactions are the core of what Firefly III is so this will mean changes. This is what the database will look like:

![transaction-group-basic](https://user-images.githubusercontent.com/5889984/54881773-d6880000-4e53-11e9-878e-79c2b6a63265.png)

And this is what a split transaction is going to look like:

![new-split-group](https://user-images.githubusercontent.com/5889984/54881803-2961b780-4e54-11e9-9efa-c62f55dc7736.png)

Everything in Firefly III is going to have to be rewritten to adhere to some basic constraints:

* The code, the UI and the API will only refer to the ID of the **TransactionGroup**. The ID's of the **TransactionJournal**s and the **Transaction**s will be hidden as much as possible.
* Budget, bill, category and tags will be linked to the **TransactionJournal**, and to no other objects. This means that several feature requests can be fulfilled _and_ the code will be easier too.
* The **Transaction**s will hold currency and amount information. One is positive, one is negative.
* The **TransactionJournal** will always have 2 **Transaction**s. One positive, one negative.

As far as most users are concerned nothing will change although the ID's of their transactions will appear to have been updated. Already, I have written several upgrade routines that convert from the old style to the new style.

During the next few weeks (this will take some time) I am going to revamp the API and the documentation, as well as the user interface to be able to cope with this.

New API version will be **0.10.0**, with documentation coming soon. I will set up a test instance on https://dev.firefly-iii.org (it's not live yet).

Thanks for your time.

Cheers,
James

**Comment by JC5** [comment]
I forgot to mention: just subscribe to this issue and I'll keep you posted on the progress.

**Comment by vitobotta** [comment]
So we won't notice any issue upon upgrading anyway or did I misunderstand? Thanks

**Comment by JC5** [comment]
No you should not. But once the dev instance is up I would appreciate if some people tried upgrading their current database to see if it works properly. 

The API will change and documentation for it will also follow ASAP.

**Comment by vitobotta** [comment]
I can upgrade as soon as it's up and let you know if I see any issues. I don't have much data in my database yet, though.

**Comment by JC5** [comment]
Here's an update guys. Sorry it took so long.

https://www.patreon.com/posts/26583431

**Comment by emansih** [comment]
![Screenshot (93)](https://user-images.githubusercontent.com/5896415/57195588-124fd280-6f87-11e9-8342-51b1912cfe36.png)


:(

**Comment by JC5** [comment]
Yes, this will in fact happen on most pages :grin:

I had to rebuild all transaction query code and all lists from the ground up, both for the API and for the normal site. Instead of upgrading the classes involved, I decided to make them error out. Whenever they're called, they'll intentionally crash with something like "I am deprecated" or "Do not use me".

The disadvantage is that 90% of the [dev.firefly-iii.org](https://dev.firefly-iii.org/) website doesn't work. But it shows me what to fix next.

Two important pages work: showing a single transaction and creating a new transaction (at least: showing the form works).
